lib/Sanger/CGP/NgsQc/ValidateSampleMetaAndBamHeader.pm

```diff
@@ -0,0 +1,755 @@
+    chomp(@in = <$FH>);
+    close $FH;
+  }
+  return \@in;
+}
+
+sub validate_samples {
+  my ($opts) = @_;
+  my $in_array;
+  if (! -e $opts->{'i'}) {
+    die ("--------\nError: Can not find input $opts->{'i'}\n--------\n");
+  }
+  $in_array = input_to_array($opts->{'i'}, $opts->{'f'});
+
+  my @lines = @$in_array;
+
+  my ($donor_index, $tissue_index, $is_normal_index, $sample_index, $bam_index);
+  my @out;
+
+  ### check input table headers
+  my $header_line = '';
+
+  if ($lines[0] =~ m/^\#|^DONOR_ID/i) {
+    warn "--------\nTake the first line as header line.\n";
+    $header_line = $lines[0]; shift @lines;
+    # assign column index
+    my @colnames = split("\t", $header_line);
+    my %header_checks;
+    my @temp = split(" ", $HEADER_IS_NORMAL); my $header_is_normal_match = $temp[0];
+    for (my $i = 0; $i < scalar @colnames; $i++) {
+      if ($colnames[$i] =~ m/$HEADER_DONOR_ID/i) {
+        $donor_index = $i;
+        if (! exists $header_checks{$HEADER_DONOR_ID}) {
+          warn "Column $colnames[$i]:$donor_index -> $HEADER_DONOR_ID\n";
+          $header_checks{$HEADER_DONOR_ID} = 1;
+        } else {
+          die("--------\nError: Duplicated header $HEADER_DONOR_ID found!\n--------\n");
+        }
+      } elsif ($colnames[$i] =~ m/$HEADER_TISSUE_ID/i) {
+        $tissue_index = $i;
+        warn "Column $colnames[$i]:$tissue_index -> $HEADER_TISSUE_ID\n";
+        if (! exists $header_checks{$HEADER_TISSUE_ID}) {
+          $header_checks{$HEADER_TISSUE_ID} = 1;
+        } else {
+          die("--------\nError: Duplicated header $HEADER_TISSUE_ID found!\n--------\n");
+        }
+      } elsif ($colnames[$i] =~ m/$header_is_normal_match/i) {
+        $is_normal_index = $i;
+        warn "Column $colnames[$i]:$is_normal_index -> $HEADER_IS_NORMAL\n";
+        if (! exists $header_checks{$HEADER_IS_NORMAL}) {
+          $header_checks{$HEADER_IS_NORMAL} = 1;
+        } else {
+          die("--------\nError: Duplicated header $HEADER_IS_NORMAL found!\n--------\n");
+        }
+      } elsif ($colnames[$i] =~ m/$HEADER_SAMPLE_ID/i) {
+        $sample_index = $i;
+        warn "Column $colnames[$i]:$sample_index -> $HEADER_SAMPLE_ID\n";
+        if (! exists $header_checks{$HEADER_SAMPLE_ID}) {
+          $header_checks{$HEADER_SAMPLE_ID} = 1;
+        } else {
+          die("--------\nError: Duplicated header $HEADER_SAMPLE_ID found!\n--------\n");
+        }
+      } elsif ($colnames[$i] =~ m/$HEADER_BAM/i) {
+        $bam_index = $i;
+        warn "Column $colnames[$i]:$bam_index -> $HEADER_BAM\n";
+        if (! exists $header_checks{$HEADER_BAM}) {
+          $header_checks{$HEADER_BAM} = 1;
+        } else {
+          die("--------\nError: Duplicated header $HEADER_BAM found!\n--------\n");
+        }
+      }
+    }
+
+    if (!exists $header_checks{$HEADER_DONOR_ID}) {die("--------\nError: $HEADER_DONOR_ID column is missing!\n--------\n");};
```

Comment:
Changed die to croak, some to confess
left warn unchanged, because they are mainly to print progress and status